### DISCUSSION TITLE
Custom nullability check for kotlin and other @NotNull types

### DISCUSSION THREAD

**Discussion by IceBlizz6** [hibernate/hibernate-orm#6837] [Design Proposals]
I'm currently working on a Kotlin codebase that is using Hibernate.
In kotlin it is possible to check for nullability through reflection.
![image](https://github.com/hibernate/hibernate-orm/assets/38980218/516e7ab7-f92e-4d58-a1a0-a412ed5afaf4)

I want a way to tell Hibernate to use this in order to apply nullability.
This will affect schema generation (not null to columns) and validation,
similar to how jakarta.validation.constraints.NotNull annotation works now.

My suggestion is to give us a way to add or override the existing nullability check.
The simplest example of how this could look like is:

=== Hibernate codebase ===
```
interface NullabilityChecker {
    // nullable return value so that we can return null if indeterminate, if so it will fall back to the default behavior
    public isNullable(field: Field): Boolean
}
```

=== My project ===
```
class KotlinNullabilityChecker : NullabilityChecker {
    override fun isNullable(field: Field): Boolean {
        return field.kotlinProperty!!.returnType.isMarkedNullable
    }
}

...

map[AvailableSettings.NULLABILITY_CHECKER] = "org.example.KotlinNullabilityChecker"
```

In this way it will be possible to easily configure nullability check for kotlin projects,
And it will also make it easier to check for other NotNull annotations.

**Comment by beikov** [comment]
You can implement a custom `Integrator`, similar to how it works for bean validation annotations as you can see in `BeanValidationIntegrator`. I don't think we need a new SPI for this. The integrator is good enough and allows for even more features.

**Comment by IceBlizz6** [comment]
Thank you @beikov  !

I solved it using `Integrator`
Here is my solution:

```
class KotlinNullabilitySupport : Integrator {
	override fun integrate(metadata: Metadata, bootstrapContext: BootstrapContext, sessionFactory: SessionFactoryImplementor) {
		metadata.entityBindings.forEach { assignNullability(it) }
	}

	private fun assignNullability(entityBinding: PersistentClass) {
		val entityName = entityBinding.entityName
		val entityProperties = entityBinding.properties
		val kotlinProperties = Class.forName(entityName)
			.kotlin
			.declaredMemberProperties
		entityProperties.forEach { entityProperty ->
			val property = kotlinProperties.single { kotlinProperty -> kotlinProperty.name == entityProperty.name }
			val isNullable = property.returnType.isMarkedNullable
			entityProperty.resetOptional(isNullable)
		}
	}

	override fun disintegrate(
		sessionFactory: SessionFactoryImplementor,
		serviceRegistry: SessionFactoryServiceRegistry
	) = Unit
}
```
